### DISCUSSION TITLE
PlatformNotSupportedException thrown when Service has Begin and End Async methods

### DISCUSSION THREAD

**Discussion by LijoAlex87** [CoreWCF/CoreWCF#1086] [General]
I have a service in WCF which I am migrating to CoreWCF. It has async methods defined in contract like below

```
[ServiceContract]
    public interface IEchoService
    {

        [OperationContract(Action = "*", ReplyAction = "*", AsyncPattern = true)]
        [XmlSerializerFormat]
        IAsyncResult BeginDummyOperation(string soapRequest, AsyncCallback asyncCallback, object asyncState);

      DummyOperationResponse EndDummyOperation(IAsyncResult result);


    }

```

I setup the service as below

```
app.UseServiceModel(builder =>
            {
                builder.AddService<EchoService>((serviceOptions) => { })
                .AddServiceEndpoint<EchoService, IEchoService>(new BasicHttpBinding(), "/EchoService/basichttp")
                .AddServiceEndpoint<EchoService, IEchoService>(myWSHttpBinding, "/EchoService/WSHttps");
            });
```

When I run the service, I get PlatformNotSupportedException. Is AsyncResult not supported on service side or am I missing something ? Any insights would be super helpful 

```
Unhandled exception. System.PlatformNotSupportedException: Operation is not supported on this platform.
   at CoreWCF.Dispatcher.OperationInvokerBehavior.CoreWCF.Description.IOperationBehavior.ApplyDispatchBehavior(OperationDescription description, DispatchOperation dispatch)
   at CoreWCF.Description.DispatcherBuilder.BindOperations(ContractDescription contract, ClientRuntime proxy, DispatchRuntime dispatch)
   at CoreWCF.Description.DispatcherBuilder.InitializeServiceHost(ServiceHostBase serviceHost, IServiceProvider services)
   at CoreWCF.Description.DispatcherBuilder.BuildDispatcher[TService](ServiceConfiguration`1 serviceConfig, IServiceProvider services)
   at CoreWCF.Configuration.ServiceConfiguration`1.GetDispatchers()
   at CoreWCF.Configuration.DispatcherBuilderImpl.BuildDispatchers(Type serviceType)
   at CoreWCF.Channels.ServiceModelHttpMiddleware.BuildBranch()
   at CoreWCF.Channels.ServiceModelHttpMiddleware.EnsureBranchBuilt()
   at CoreWCF.Channels.ServiceModelHttpMiddleware.ServiceBuilderOpenedCallback(Object sender, EventArgs e)
   at CoreWCF.Channels.CommunicationObject.OnOpened()
--- End of stack trace from previous location ---
   at CoreWCF.Configuration.WrappingIServer.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Microsoft.AspNetCore.Builder.WebApplication.Run(String url)
```

**Comment by LijoAlex87** [comment]
@g7ed6e Do you have any insights on this as well :) .  Thanks for all the help

**Comment by LijoAlex87** [comment]
From the code here https://github.com/CoreWCF/CoreWCF/blob/8aad8508d82a65bc6ca79178943321fa9bf70f24/src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/OperationInvokerBehavior.cs#L60 looks like this is no longer supported. Is that the case ?

Or Should I be replacing the OprationInvokerBehaviour with some other behaviour for Async methods. Please let me know

```
 if (description.BeginMethod != null)
                {
                    // only async method is present on the contract
                    throw new PlatformNotSupportedException();
                    //dispatch.Invoker = new AsyncMethodInvoker(description.BeginMethod, description.EndMethod);
                }
```

**Reply by mconnew** [reply]
Looks like it's not supported. I forgot I had done that, but I remember now. The code around this is complicated and the APM style of calls has basically been deprecated. This shouldn't be a big deal though as you can still bring your implementation over with a minor change. Change your contract like this:
```c#
    [ServiceContract]
    public interface IEchoService
    {
        [OperationContract(Action = "*", ReplyAction = "*")]
        [XmlSerializerFormat]
        Task<DummyOperationResponse> DummyOperation(string soapRequest);
    }
```
And add this method to your implementation:
```c#
    public class EchoService : IEchoService
    {
        Task<DummyOperationResponse> DummyOperation(string soapRequest)
        {
            return Task.Factory.FromAsync(BeginDummyOperation, EndDummyOperation, soapRequest, null);
        }

        public IAsyncResult BeginDummyOperation(string soapRequest, AsyncCallback asyncCallback, object asyncState)
        {
            // Existing implementation
        }

        public DummyOperationResponse EndDummyOperation(IAsyncResult result)
        {
            // Existing implemenation
        }
}
```

The only time this won't work is if you have out parameters on the EndXXXX method. I've never seen this "in the wild" and only seen it used on internal methods, but it's technically possible. Then things get complicated as you need to switch to using message contracts and have an incoming request object, and a response object. It looks like you are kind of doing that today with your DummyOperationResponse, but with message contracts it allows you to have a client with an out parameter when using sync or APM style contracts.  
I would much prefer adding support for named tuple return values than add support for APM though. E.g. ` Task<(string result, bool success)> DoWorkAsync();` would be the equivalent of `string DoWork(out bool success);` or `IAsyncResult BeginDoWork(AsyncCallback asyncCallback, object syncState); string EndDoWork(IAsyncResult result, out bool success);`

**Reply by LijoAlex87** [reply]
@mconnew If we switch contract to Task<DummyOperationResponse> DummyOperation(string soapRequest) will existing callers of the service face any issue ? Will they need to make any changes at their end to call the service with updated contarct now

**Reply by LijoAlex87** [reply]
Also, I tried changing it as you suggested

```
public IAsyncResult BeginDummyOperation(string soapRequest, AsyncCallback asyncCallback, object asyncState)
        {
            // Existing implementation
        }

        public DummyOperationResponse EndDummyOperation(IAsyncResult result)
        {
            // Existing implementation
        }

        public Task<DummyOperationResponse> DummyOperation(string soapRequest)
        {
            return Task.Factory.FromAsync<DummyOperationResponse>(BeginDummyOperation, EndDummyOperation, soapRequest, TaskCreationOptions.None);
        }
```

But this was giving below compilation errors 

 error CS1503: Argument 1: cannot convert from 'method group' to 'Func<DummyOperationResponse, AsyncCallback, object, IAsyncResult>'
error CS0407: 'DummyOperationResponse EWSService.EndDummyOperation(IAsyncResult)' has the wrong return type
 error CS1503: Argument 3: cannot convert from 'string' to 'Microsoft.Exchange.Services.Wcf.DummyOperationResponse'

as FromAsync is of following declaration
        public Task FromAsync<TArg1>(Func<TArg1, AsyncCallback, object?, IAsyncResult> beginMethod, Action<IAsyncResult> endMethod, TArg1 arg1, object? state, TaskCreationOptions creationOptions);

**Reply by mconnew** [reply]
You have passed `TaskCreationOptions.None` instead of `null`. The `null` value is the optional state value to associate with the `Task`. `Task` implements `IAsyncResult` and the passed in state is what you would get if you called `((IAsyncResult)task).AsyncState`. For the creation options, `TaskCreationOptions.None` is the default so you don't need to pass it. If you want to explicitly pass this, then you need to include the null before the `TaskCreationOptions`. You need to do one of these:
```c#
        Task<DummyOperationResponse> DummyOperation(string soapRequest)
        {
            return Task.Factory.FromAsync(BeginDummyOperation, EndDummyOperation, soapRequest, null);
        }
```
Or
```c#
        Task<DummyOperationResponse> DummyOperation(string soapRequest)
        {
            return Task.Factory.FromAsync(BeginDummyOperation, EndDummyOperation, soapRequest, null, TaskCreationOptions.None);
        }
```

**Reply by LijoAlex87** [reply]
Thanks a lot @mconnew for the details. I was able to make the change following your suggestion

If we switch contract to Task<DummyOperationResponse> DummyOperation(string soapRequest)  from the current Begin* and End* form will existing callers of the service face any issue ? Will they need to make any changes at their end to call the service with updated contarct now

**Reply by mconnew** [reply]
No , the async nature of the operation is irrelevant when sent across the network. A client can call APM, Sync, or Task and the server can be any one of the three (in wcf) or the two for CoreWCF.

**Reply by LijoAlex87** [reply]
@mconnew We also have some contracts where we have operations supporting WebGet like below

```
[OperationContract(AsyncPattern = true)]
[JsonRequestFormat(Format = JsonRequestFormat.Custom)]
[WebGet]
IAsyncResult BeginGetDummyOperation(string param1, string param2, int param3, AsyncCallback callback, object state);

Stream EndGetDummyOperation(IAsyncResult result);
```
Here also can I do like above. Just double checking this holds true for WebGet scenarios as well

Replace Begin* and End* with Task method in Interface as shown below

```
[OperationContract]
[JsonRequestFormat(Format = JsonRequestFormat.Custom)]
[WebGet]
Task<Stream> GetDummyOperation(string param1, string param2, int param3);
```

And correspoding implementation

```
public Task<Stream> GetDummyOperation(string param1, string param2, int param3)
{
    return Task.Factory.FromAsync(BeginGetPeopleICommunicateWith, EndGetPeopleICommunicateWith, param1, param2, param3, null);
}
```

Let me know if there is any issue if we do this approach

**Reply by mconnew** [reply]
I've never directly tried that exact scenario but I would expect it to work. If it doesn't, I would consider it a bug.
